pkg/compactor/blocks_cleaner_test.go

```diff
@@ -406,6 +410,231 @@ func TestBlocksCleaner_ShouldRemoveMetricsForTenantsNotBelongingAnymoreToTheShar
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), time.Hour)
+	assert.ElementsMatch(t, []ulid.ULID{}, result.GetULIDs())
+
+	// Normal operation - varying retention period.
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 4*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{}, result.GetULIDs())
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 3*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{id1}, result.GetULIDs())
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 2*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{id1, id2}, result.GetULIDs())
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 1*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{id1, id2, id3}, result.GetULIDs())
+
+	// Avoiding redundant marking - blocks already marked for deletion.
+
+	mark1 := &bucketindex.BlockDeletionMark{ID: id1}
+	mark2 := &bucketindex.BlockDeletionMark{ID: id2}
+
+	idx.BlockDeletionMarks = bucketindex.BlockDeletionMarks{mark1}
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 3*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{}, result.GetULIDs())
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 2*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{id2}, result.GetULIDs())
+
+	idx.BlockDeletionMarks = bucketindex.BlockDeletionMarks{mark1, mark2}
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 3*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{}, result.GetULIDs())
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 2*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{}, result.GetULIDs())
+
+	result = listBlocksOutsideRetentionPeriod(idx, time.Unix(10, 0), 1*time.Second)
+	assert.ElementsMatch(t, []ulid.ULID{id3}, result.GetULIDs())
+}
+
+func TestBlocksCleaner_ShouldRemoveBlocksOutsideRetentionPeriod(t *testing.T) {
+	bucketClient, _ := cortex_testutil.PrepareFilesystemBucket(t)
+	bucketClient = bucketindex.BucketWithGlobalMarkers(bucketClient)
+
+	ts := func(hours int) int64 {
+		return time.Now().Add(time.Duration(hours)*time.Hour).Unix() * 1000
+	}
+
+	block1 := createTSDBBlock(t, bucketClient, "user-1", ts(-10), ts(-8), nil)
+	block2 := createTSDBBlock(t, bucketClient, "user-1", ts(-8), ts(-6), nil)
+	block3 := createTSDBBlock(t, bucketClient, "user-2", ts(-10), ts(-8), nil)
+	block4 := createTSDBBlock(t, bucketClient, "user-2", ts(-8), ts(-6), nil)
+
+	cfg := BlocksCleanerConfig{
+		DeletionDelay:      time.Hour,
+		CleanupInterval:    time.Minute,
+		CleanupConcurrency: 1,
+	}
+
+	ctx := context.Background()
+	logger := log.NewNopLogger()
+	reg := prometheus.NewPedanticRegistry()
+	scanner := tsdb.NewUsersScanner(bucketClient, tsdb.AllUsers, logger)
+	cfgProvider := newMockConfigProvider()
+
+	cleaner := NewBlocksCleaner(cfg, bucketClient, scanner, cfgProvider, logger, reg)
+
+	assertBlocksExist := func(exists1, exists2, exists3, exists4 bool) {
```

Comment:
Mixed feeling about this function. Maybe it's easier to read when you call it, if you change it to pick in input only 1 block, and you call `assertBlockExists(block meta, exists bool)` multiple times, one per block

Reply:
I agree, the meaning of each true/false is not clear. A nice alternative would be to list the blocks explicitly:

```
assertBlocksExist(block1, block2, block3)
```